File: src/2025-03-04-12-20-21--deepseek-vr-0304/python/exercises/practice/variable-length-quantity/variable_length_quantity.py
def encode(numbers):
    encoded = []
    for number in numbers:
        if number < 0:
            raise ValueError("value out of range")
        chunks = []
        if number == 0:
            chunks.append(0)
        else:
            while number > 0:
                chunks.append(number & 0x7F)
                number >>= 7
            chunks.reverse()
            for i in range(len(chunks) - 1):
                chunks[i] |= 0x80
        encoded.extend(chunks)
    return encoded



def decode(bytes_):
    result = []
    current_value = 0
    in_sequence = False
    for byte in bytes_:
        if byte < 0 or byte > 255:
            raise ValueError("invalid byte value")
        current_value = (current_value << 7) | (byte & 0x7F)
        if byte & 0x80:
            in_sequence = True
        else:
            result.append(current_value)
            current_value = 0
            in_sequence = False
    if in_sequence:
        raise ValueError("incomplete sequence")
    return result

File: src/2025-03-04-12-20-21--deepseek-vr-0304/go/exercises/practice/variable-length-quantity/variable_length_quantity.go
package variablelengthquantity

import "errors"

func EncodeVarint(input []uint32) []byte {
	var result []byte
	for _, num := range input {
		var chunks []byte
		n := num
		if n == 0 {
			result = append(result, 0)
			continue
		}
		for n > 0 {
			chunks = append(chunks, byte(n%128))
			n = n / 128
		}
		// Reverse chunks
		for i, j := 0, len(chunks)-1; i < j; i, j = i+1, j-1 {
			chunks[i], chunks[j] = chunks[j], chunks[i]
		}
		// Set high bit on all but last chunk
		for i := 0; i < len(chunks)-1; i++ {
			chunks[i] |= 0x80
		}
		result = append(result, chunks...)
	}
	return result
}

func DecodeVarint(input []byte) ([]uint32, error) {
	var result []uint32
	var current uint32
	for _, b := range input {
		current = current<<7 | uint32(b&0x7F)
		if (b & 0x80) == 0 {
			result = append(result, current)
			current = 0
		}
	}
	if current != 0 {
		return nil, errors.New("incomplete sequence")
	}
	return result, nil
}
